app/routes.py

```diff
@@ -1,2 +1,274 @@
+                task_response = {"details": "Invalid data"}
+                return make_response(task_response, 400)
+
+        new_task = Task(title=request_body["title"],
+                        description=request_body["description"],
+                        completed_at=request_body["completed_at"])
+
+        db.session.add(new_task)
+        db.session.commit()
+        if not request_body["completed_at"]:
+            is_complete = False
+        else:
+            is_complete = True
+        task_response = {"task":
+                         {"id": new_task.task_id,
+                          "title": new_task.title,
+                          "description": new_task.description,
+                          "is_complete": is_complete}
+                         }
+        return make_response(task_response, 201)
+
+
+@tasks_bp.route("/<task_id>", methods=["GET", "PUT", "DELETE"])
+def handle_task(task_id):
+    task = Task.query.get(task_id)
+    if not task:
+        return make_response(f"Task #{task_id} Not Found", 404)
+
+    if request.method == "GET":
+        if not task.completed_at:
+            is_complete = False
+        else:
+            is_complete = True
+        task_response = {"task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": is_complete,
+        }}
+        if task.goal_id:
+            task_response["task"]["goal_id"] = task.goal_id
+        return make_response(task_response, 200)
+    elif request.method == "PUT":
+        form_data = request.get_json()
+        if ("completed_at" not in form_data or
+                form_data["completed_at"] == None):
+            is_complete = False
+        else:
+            is_complete = True
+        task.title = form_data["title"]
+        task.description = form_data["description"]
+        task.completed_at = form_data["completed_at"]
+        db.session.commit()
+        task_response = {"task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": is_complete
+        }}
+        return make_response(task_response, 200)
+
+    elif request.method == "DELETE":
+        db.session.delete(task)
+        db.session.commit()
+        task_response = {
+            "details": f"Task {task.task_id} \"{task.title}\" successfully deleted"
+        }
+        return make_response(task_response, 200)
+
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def handle_task_complete(task_id):
+    task = Task.query.get(task_id)
+    if not task:
+        return make_response(f"Task #{task_id} Not Found", 404)
+
+    task.completed_at = datetime.datetime.utcnow()
+    db.session.commit()
+    channel = "channel=task-notifications"
+    text = f"text=Someone just completed the task {task.title}"
+
+    path = f"https://slack.com/api/chat.postMessage?{channel}&{text}"
```

Comment:
Rather than us being responsible for adding query params into the url directly, we can make use of the `params` named param on the call to requests.post. Doing so would allow the requests library itself to do any necessary encoding of our data values, and we get to work with a nice Python dictionary.

Even better, since this is a post request, we can send the parameters in the post body using the  `data` named parameter (similar to when we pass things as `json` but the way the values are packed into the request is a little different).